reconciler/reconciler_test.go

```diff
@@ -267,40 +264,29 @@ func testReconciler(t *testing.T, batchOps bool) {
 			h.expectNotFound(ID_1)
 			h.expectNotFound(ID_2)
 			h.expectNotFound(ID_3)
-			h.triggerFullReconciliation()
+			h.triggerPrune()
 			h.expectOps(opDelete(1), opDelete(2), opDelete(3), opPrune(0))
 			h.waitForReconciliation()
 
-		}
-	})
-
-	// Validate that the metrics are populated and make some sense.
-	getInt := func(v expvar.Var) int64 {
-		if v, ok := v.(*expvar.Int); ok && v != nil {
-			return v.Value()
-		}
-		return -1
-	}
+			// Validate metrics.
+			assert.Greater(t, getInt(h.m.PruneCountVar.Get("test")), int64(0), "PruneCount")
+			assert.Greater(t, getFloat(h.m.PruneDurationVar.Get("test")), float64(0), "PruneDuration")
+			assert.Equal(t, getInt(h.m.PruneCurrentErrorsVar.Get("test")), int64(0), "PruneCurrentErrors")
 
-	getFloat := func(v expvar.Var) float64 {
-		if v, ok := v.(*expvar.Float); ok && v != nil {
-			return v.Value()
 		}
-		return -1
-	}
+	})
 
-	assert.Greater(t, getInt(expVarMetrics.FullReconciliationCountVar.Get("test")), int64(0), "FullReconciliationCount")
-	assert.Greater(t, getFloat(expVarMetrics.FullReconciliationDurationVar.Get("test/prune")), float64(0), "FullReconciliationDuration/prune")
-	assert.Greater(t, getFloat(expVarMetrics.FullReconciliationDurationVar.Get("test/update")), float64(0), "FullReconciliationDuration/update")
-	assert.Equal(t, getInt(expVarMetrics.FullReconciliationCurrentErrorsVar.Get("test")), int64(0), "FullReconciliationCurrentErrors")
+	runTest("refreshing", 100*time.Millisecond /* refresh interval */, func(h testHelper) {
+		t.Logf("Inserting test objects 1, 2 & 3")
+		h.insert(ID_1, NonFaulty, reconciler.StatusPending())
+		h.expectOp(opUpdate(ID_1))
+		h.expectStatus(ID_1, reconciler.StatusKindDone, "")
 
-	assert.Greater(t, getInt(expVarMetrics.IncrementalReconciliationCountVar.Get("test")), int64(0), "IncrementalReconciliationCount")
-	assert.Greater(t, getFloat(expVarMetrics.IncrementalReconciliationDurationVar.Get("test/update")), float64(0), "IncrementalReconciliationDuration/update")
-	assert.Greater(t, getFloat(expVarMetrics.IncrementalReconciliationDurationVar.Get("test/delete")), float64(0), "IncrementalReconciliationDuration/delete")
-	assert.Greater(t, getInt(expVarMetrics.IncrementalReconciliationTotalErrorsVar.Get("test")), int64(0), "IncrementalReconciliationTotalErrors")
-	assert.Equal(t, getInt(expVarMetrics.IncrementalReconciliationCurrentErrorsVar.Get("test")), int64(0), "IncrementalReconciliationCurrentErrors")
+		// With 1 second delay and 100ms refresh interval, we should see at least 5 refreshes.
+		time.Sleep(time.Second)
```

Comment:
Not great to have this sort of approximate and time-based test case. Suggestions on how do this better are welcome.
Validated with `go test -c -race` and `stress ./reconciler.test` and got `1m50s: 575 runs so far, 0 failures` so it shouldn't at least be flaky.

Reply:
I'm a big fan of using FakeClock, something like that:
https://github.com/cilium/cilium/pull/32588/files#diff-35fa8fe44daee2f4803801edfd6b2533b74caf8958ed58b3763feca25919e5a5R27

Reply:
Thanks, I'll look into that!

Reply:
Rewrote the test to not delay.